Implement Metric Server

simon-mo opened the pull request:
Implement the metric server described in #423. This will solves issue #401 and #412. 

Note:
- No test is needed because we replaced the original model container metrics with the new design. That's sufficient for testing the new implementation.
- Extensive API Docs are added for client facing functions.

simon-mo committed:
Fix dict unpacking lint error

simon-mo committed:
Format Code

simon-mo committed:
Update yapf

simon-mo committed:
Update yapf and reformat code

simon-mo committed:
Check yapf version in shell script

simon-mo committed:
Change `echo` to `cat`

simon-mo committed:
Add python version check; almost ready

simon-mo committed:
Format code with python version 2.7.12

simon-mo committed:
Merge branch 'develop' of https://github.com/ucbrise/clipper into develop

simon-mo committed:
Merge branch 'develop' of https://github.com/ucbrise/clipper into develop

simon-mo committed:
Merge branch 'develop' of https://github.com/ucbrise/clipper into develop

simon-mo committed:
Merge branch 'develop' of https://github.com/ucbrise/clipper into develop

simon-mo committed:
Merge branch 'develop' of https://github.com/ucbrise/clipper into develop

simon-mo committed:
Merge branch 'develop' of https://github.com/ucbrise/clipper into develop

simon-mo committed:
Merge branch 'develop' of https://github.com/ucbrise/clipper into develop

simon-mo committed:
Modify RPCDockerfile to install redis+other dependencies

simon-mo committed:
Add metric subpackage; refactor rpc.py metric

simon-mo committed:
Add metric pool in client

simon-mo committed:
Install redis-server via apt-get

simon-mo committed:
Change api version to clipper version

simon-mo committed:
Add user defined metric example

simon-mo committed:
Add docs

simon-mo committed:
Format code

simon-mo committed:
Merge branch 'develop' of https://github.com/ucbrise/clipper into metric_server

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1135/
Test FAILed.

simon-mo committed:
Use pip install conda to install cloudpickle

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1136/
Test FAILed.

simon-mo committed:
Add redis to requirements, setup and test dockerfile

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1137/
Test FAILed.

simon-mo committed:
Change all conda install cloudpickle to pip

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1138/
Test FAILed.

simon-mo commented:
jenkins test this please

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1141/
Test FAILed.

dcrankshaw committed:
update conda versions

dcrankshaw committed:
consolidate cloudpickle install and update anaconda

simon-mo committed:
Revert "Change all conda install cloudpickle to pip"

This reverts commit 042ebb27bc7e615de32827165cd06f8c440e3887.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1146/
Test FAILed.

simon-mo committed:
Add prom_client to test dockerfile

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1149/
Test FAILed.

simon-mo commented:
I just added another package and conda seg fault again?! 😟

simon-mo commented:
jenkins test this please

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1150/
Test FAILed.

simon-mo committed:
Fix path finding for metric config

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1153/
Test FAILed.

simon-mo commented:
jenkins test this please

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1154/
Test FAILed.

simon-mo committed:
Try longer timeout

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1155/
Test FAILed.

simon-mo committed:
Add collect_metric flag in RPCService

- Since we might not be collection etrics in RPCService, we default
the starting of redis service to metric packge init

simon-mo committed:
Remove unnecessary start_redis_daemon

- we started it in import already

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1158/
Test FAILed.

simon-mo commented:
jenkins test this please

simon-mo commented:
^seg-fault again

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1159/
Test FAILed.

simon-mo committed:
Add collect_metric flag to Server as well

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1160/
Test FAILed.

simon-mo commented:
jenkins test this please

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1161/
Test FAILed.

simon-mo committed:
Update conda in RPCDockerFile

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1163/
Test FAILed.

simon-mo committed:
Try 5.0.1 anaconda

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1164/
Test FAILed.

simon-mo committed:
Revert "Try 5.0.1 anaconda"

This reverts commit 525dcd5fd4f78ee25a3200b4641a8866d4509223.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1165/
Test PASSed.

dcrankshaw commented on clipper_admin/clipper_admin/__init__.py:
```diff
@@ -6,3 +6,4 @@
 from . import deployers
 from .version import __version__
 from .exceptions import ClipperException, UnconnectedException
+from . import metric
```
Change `metric` to `metrics`?

dcrankshaw commented on clipper_admin/clipper_admin/metric/__init__.py:
```diff
@@ -0,0 +1,5 @@
+from . import server
+
+if not server.redis_daemon_exist():
+    server.start_redis_daemon()
```
Will this start a Redis daemon any time `clipper_admin` is imported?

simon-mo replied:
You are correct. I just deleted `from . import metrics` in `__init__.py`. 

Now only when users type `import clipper_admin.metrics as metrics`, the redis server will be started (which is the expected behavior)

dcrankshaw commented on clipper_admin/clipper_admin/metric/client.py:
```diff
@@ -0,0 +1,82 @@
+    if name in metric_pool:
+        return
+
+    messege_dict = {
```
typo: s/messege/message

dcrankshaw commented on clipper_admin/clipper_admin/metric/client.py:
```diff
@@ -0,0 +1,82 @@
+        The name of metric to be reported. The name must the name of metric that 
+        has been added via `add_metric`.
+    val: int or float
+        The value of metric to be reported. The value must be able to be casted 
```
change the second sentence to "...must be able to be cast to float."

dcrankshaw commented on clipper_admin/clipper_admin/metric/client.py:
```diff
@@ -0,0 +1,82 @@
+        The value of metric to be reported. The value must be able to be casted 
+        as float. 
+    """
+    assert name in metric_pool, "{} has not been added. \
```
Instead of using an assert, can you raise an exception?

dcrankshaw commented on clipper_admin/clipper_admin/metric/client.py:
```diff
@@ -0,0 +1,82 @@
+        If it's not specificed and the metric_type is 'Histogram', the default buckets
+        [5, 10, 20, 35, 50, 75, 100, 150, 200, 250, 300, 400, 500] will be used. 
+    """
+    assert metric_type in set(
```
Change assert to raising an exception

dcrankshaw commented on clipper_admin/clipper_admin/metric/server.py:
```diff
@@ -0,0 +1,141 @@
+            name, metric_pool.keys()))
+
+
+def handle_messege(messege_dict, metric_pool):
```
s/messege/message

dcrankshaw commented on clipper_admin/clipper_admin/metric/server.py:
```diff
@@ -0,0 +1,141 @@
+
+def handle_messege(messege_dict, metric_pool):
+    """
+    Handle a messege dictionary, dispatch request to add or report call
```
s/messege/message

dcrankshaw commented on clipper_admin/clipper_admin/metric/server.py:
```diff
@@ -0,0 +1,141 @@
+    sub.subscribe(CHANNEL_NAME)
+
+    metric_pool = {}
+    for messege in sub.listen():  # Blocking, will run forever
```
s/messege/message

dcrankshaw commented on clipper_admin/clipper_admin/metric/server.py:
```diff
@@ -0,0 +1,141 @@
+        try:
+            messege_dict = json.loads(messege['data'])
+            validate_schema(messege_dict)
+            handle_messege(messege_dict, metric_pool)
```
s/messege/message

dcrankshaw commented on clipper_admin/clipper_admin/metric/server.py:
```diff
@@ -0,0 +1,141 @@
+            messege_dict = json.loads(messege['data'])
+            validate_schema(messege_dict)
+            handle_messege(messege_dict, metric_pool)
+        except Exception as e:
```
Do you want to catch all exceptions here? Or would it make sense to be more specific about which exceptions to catch?

dcrankshaw reviewed (changes requested):
This is awesome!  I just have a couple minor comments.

Also, how hard would it be to add this to the model containers in other languages (e.g. Java)? Can you file an issue to remind us to do this?

simon-mo committed:
Address comments

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1177/
Test FAILed.

simon-mo commented:
Jenkins test this please

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1179/
Test FAILed.

simon-mo commented:
Jenkins test this please

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1191/
Test FAILed.

dcrankshaw committed:
Merge branch 'develop' into metric_server

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1202/
Test FAILed.

dcrankshaw commented:
jenkins test this please

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1218/
Test FAILed.

dcrankshaw commented:
jenkins test this please

dcrankshaw reviewed (approved):
LGTM

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1219/
Test FAILed.

dcrankshaw committed:
fixed container build issues related to conda segfaults and out of date packages

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1223/
Test FAILed.

dcrankshaw committed:
uncommented docker build

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1224/
Test FAILed.

dcrankshaw commented:
jenkins test this please

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1226/
Test FAILed.

dcrankshaw commented:
jenkins test this please

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1228/
Test FAILed.

simon-mo committed:
Merge branch 'develop' of https://github.com/ucbrise/clipper into metric_server

simon-mo committed:
Merge branch 'metric_server' of https://github.com/simon-mo/clipper into metric_server

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1256/
Test FAILed.

simon-mo committed:
Try change the model name to prevent image caching

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1257/
Test FAILed.

simon-mo committed:
Revert "Try change the model name to prevent image caching"

This reverts commit 8a2788b0908e4bb54f74794f8ee2c7cad71bd6c9.

simon-mo committed:
Add debug lines for num_replica test

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1258/
Test FAILed.

simon-mo committed:
Add psutil to test container

simon-mo committed:
Add docker stats to debug

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1259/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1260/
Test FAILed.

dcrankshaw commented on clipper_admin/clipper_admin/metrics/config.py:
```diff
@@ -0,0 +1,10 @@
+# client and server
+
+CHANNEL_NAME = 'clipper'  # redis pub-sub channel name
+API_VERSION = "0.3.0"  # Consistent with Clipper Version.
```
Can you import this from `clipper_admin.__version__`?

dcrankshaw commented on clipper_admin/clipper_admin/metrics/schema.py:
```diff
@@ -0,0 +1,81 @@
+    s = 'Summary'
+
+
+versions = ['0.3.0']
```
Import from Clipper admin version, don't hardcode version numbers

dcrankshaw commented on clipper_admin/clipper_admin/metrics/schema.py:
```diff
@@ -0,0 +1,81 @@
+}
+
+
+def validate_schema(messege_dict):
```
message is still spelled wrong here

dcrankshaw commented on containers/python/rpc.py:
```diff
@@ -10,9 +10,10 @@
 import yaml
 from collections import deque
-from multiprocessing import Pipe, Process
+from subprocess import Popen, PIPE
```
Use `subprocess32` instead of `subprocess`

dcrankshaw reviewed (changes requested):
I have no idea why the tests are failing. I'm trying to reproduce locally.

dcrankshaw commented:
jenkins test this please

simon-mo committed:
Address comments; Fix Exception Handles in py2

simon-mo committed:
Format code

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1262/
Test FAILed.

simon-mo committed:
Remove debug lines

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1265/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1266/
Test FAILed.

simon-mo committed:
Bug fix and typo fix

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1270/
Test FAILed.

simon-mo committed:
Fix kwarg

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1271/
Test FAILed.

simon-mo committed:
Install jsonschema in dev (also test) package

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1274/
Test FAILed.

simon-mo committed:
Add dependencies to DevDocker & Clipper Req

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1278/
Test PASSed.

dcrankshaw reviewed (approved):
lgtm
